torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2738,16 +2722,23 @@ def _post_forward(
         according to the sharded flattened parameter's dtype and shape.
         """
         with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):
-            if self not in self._fsdp_graph_order:
-                self._my_fsdp_idx_in_graph = len(self._fsdp_graph_order)
-                self._fsdp_graph_order.append(self)
+            params_resharded = []
             if reshard_fn is not None:
-                reshard_fn()
+                params_resharded = reshard_fn()
             if unshard_fn is not None:
                 unshard_fn()
-            # Register pre-backward hooks to unshard the flattened parameters for
-            # the gradient computation if needed
-            output = self._register_pre_backward_hooks(output, [handle.flat_param for handle in handles])
+
+            if not self._use_param_exec_order_policy:
+                if self not in self._fsdp_graph_order:
+                    self._my_fsdp_idx_in_graph = len(self._fsdp_graph_order)
+                    self._fsdp_graph_order.append(self)
+                output = self._register_pre_backward_hooks(output, [handle.flat_param for handle in handles])
+            elif params_resharded:
+                if self not in self._fsdp_graph_order:
+                    self._my_fsdp_idx_in_graph = len(self._fsdp_graph_order)
+                    self._fsdp_graph_order.append(self)
```

Comment:
Note: This will be in terms of original `nn.Module`s, while for recursive wrapping, it will be in terms of `FullyShardedDataParallel`s.

Reply:
This seems like a correctness problem without further changes since the backward prefetching assumes the elements in `self._fsdp_graph_order` are `FullyShardedDataParallel` instances.

I think we need to make training state per handle and adjust from there.